frameworks/TunedRandomForest/exec.py

```diff
@@ -64,60 +80,112 @@ def run(dataset, config):
+                    scores = cross_val_score(
+                        estimator=random_forest,
+                        X=X_train,
+                        y=y_train,
+                        scoring=metric,
+                        error_score='raise',
+                        cv=5
+                    )
+                    max_feature_scores[value].append(statistics.mean(scores))
                 except Exception as e:
-                    log.error("Failed CV scoring for max_features=%s :\n%s", max_features_value, e)
+                    log.error("Failed CV scoring for max_features=%s :\n%s", value, e)
                     log.debug("Exception:", exc_info=True)
-                    max_feature_scores.append((math.nan, max_features_value))
-            tuning_durations.append((max_features_value, cv_scoring.duration))
-
-    log.info("Tuning scores:\n%s", sorted(max_feature_scores))
-    log.info("Tuning durations:\n%s", sorted(tuning_durations))
-    _, best_max_features_value = max(max_feature_scores) if len(max_feature_scores) > 0 else (math.nan, 'auto')
-    log.info("Training final model with `max_features={}`.".format(best_max_features_value))
-    rf = estimator(n_jobs=n_jobs,
-                   random_state=config.seed,
-                   max_features=best_max_features_value,
-                   **training_params)
-    with Timer() as training:
-        rf.fit(X_train, y_train)
+                    max_feature_scores[value].append(math.nan)
+                    break
+                finally:
+                    training_times.append(training.duration)
+                    memory_usage.append(this_process.memory_info()[0] / (2 ** 20))
+
+                if random_forest.n_estimators >= final_forest_size:
+                    log.info("Stop training because desired forest size has been reached.")
+                    break
+                if extrapolate_with_worst_case(training_times) - training_times[0] >= time_per_value:
+                    log.info("Stop training because it expects to exceed its time budget.")
+                    break
+                elif extrapolate_with_worst_case(memory_usage) >= config.max_mem_size_mb * memory_margin:
+                    log.info("Stop training because it expects to exceed its memory budget.")
+                    break
+
+                random_forest.n_estimators += step_size
+
+            tuning_durations[value] = training_times
+            memory_usage_by[value] = memory_usage
+            last_initial_fit_time = training_times[1] - training_times[0]
+
+        # TODO: Transform the dictionaries into a pandas dataframe (value, trees, score, duration, total_duration, memory, d_memory)
+        log.info("Tuning scores:\n%s", max_feature_scores)
+        log.info("Tuning durations:\n%s", tuning_durations)
+        log.info("Tuning memory:\n%s", memory_usage_by)
+
+        _, best_value = max((scores[-1], value) for value, scores in max_feature_scores.items())
```

Comment:
Add some documentation string that this retrains the best model on the entire training set?